cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,1258 @@
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <algorithm>
+#include <climits>
+#include <cmath>
+#include <cstdint>
+#include <limits>
+
+#ifndef __CUDACC__
+#  define __host__
+#  define __device__
+#  define __forceinline__
+#  include <array>
+using std::array;
+using std::max;
+using std::min;
+#else
+#  include <cuda/std/__algorithm_>
+#  include <cuda/std/array>
+using cuda::std::array;
+using cuda::std::max;
+using cuda::std::min;
+
+// disable zero checks
+#  define DISABLE_ZERO
+
+// disable nan / infinity checks
+#  define DISABLE_NANINF
+
+// jump table for indexing into data
+#  define MAX_JUMP 5
+static_assert(MAX_JUMP <= 5, "MAX_JUMP greater than max");
+
+#endif
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+namespace rfa_detail
+{
+
+template <typename Float, int LEN>
+static __device__ Float* get_shared_bin_array()
+{
+  static __shared__ Float bin_computed_array[LEN];
+  return bin_computed_array;
+}
+
+template <class T>
+struct get_vector_type
+{};
+template <>
+struct get_vector_type<double>
+{
+  using type = double2;
+};
+template <>
+struct get_vector_type<float>
+{
+  using type = float4;
+};
+
+template <class T>
+using vector_t = typename get_vector_type<T>::type;
+
+template <class T>
+std::size_t vector_size()
+{
+  return sizeof(vector_t<T>) / sizeof(T);
+}
+
+template <class T>
+inline static auto abs_max(const T&);
+
+template <>
+__host__ __device__ auto abs_max(const float4& x)
+{
+  return fmax(fmaxf(fabs(x.x), fabs(x.y)), fmax(fabs(x.z), fabs(x.w)));
+}
+
+template <>
+__host__ __device__ auto abs_max(const double4& x)
+{
+  return fmax(fmax(fabs(x.x), fabs(x.y)), fmax(fabs(x.z), fabs(x.w)));
+}
+
+template <>
+__host__ __device__ auto abs_max(const double2& x)
+{
+  return fmax(fabs(x.x), fabs(x.y));
+}
+
+template <>
+__host__ __device__ auto abs_max(const float2& x)
+{
+  return fmax(fabs(x.x), fabs(x.y));
+}
+
+template <class ftype>
+struct RFA_bins
+{
+  static constexpr auto BIN_WIDTH = std::is_same_v<ftype, double> ? 40 : 13;
+  static constexpr int MIN_EXP    = std::numeric_limits<ftype>::min_exponent;
+  static constexpr int MAX_EXP    = std::numeric_limits<ftype>::max_exponent;
+  static constexpr auto MANT_DIG  = std::numeric_limits<ftype>::digits;
+  /// Binned floating-point maximum index
+  static constexpr auto MAXINDEX = ((MAX_EXP - MIN_EXP + MANT_DIG - 1) / BIN_WIDTH) - 1;
+  // The maximum floating-point fold supported by the library
+  static constexpr auto MAXFOLD = MAXINDEX + 1;
+
+  __device__ __host__ static ftype initialize_bins(int index)
+  {
+    if (index == 0)
+    {
+      return ldexpf(0.75f, MAX_EXP);
+    }
+    else
+    {
+      return ldexpf(0.75f, MAX_EXP + MANT_DIG - BIN_WIDTH - index * BIN_WIDTH);
+    }
+  }
+};
+
+/// Class to hold a reproducible summation of the numbers passed to it
```

Comment:
We are moving towards rst style documentation, so please use `//!` instead of `///`. Applies throughout

Reply:
Is this a general rule? I wondered a few times already whether we should write done some of those guidelines.

Reply:
I am guessing I only do that where we actually need the docs for the function and rest of the comments can stay.